Fix various network-stability issues with GERTi

20kdc opened the pull request:
In particular, the following "just forward back next packet" cases have been turned into proper responses:
 ResolveAddress/AddressResolution
 RegisterNode/RegisterComplete

Running without an MNC now has a very small chance of actually working.
"ROUTE OPEN" now contains data on which route is being opened (which avoids cross-talk)

Additional note: Though I've done basic testing, I haven't got a GERTe setup, so I am not 100% sure if my changes interfere with that. They should not, but do test before merging. :)

EDIT: Apparently the "run without MNC" code isn't working properly. (Which means it's working about as well as it did before, to be honest.)

20kdc committed:
Fix various network-stability issues with GERTi

In particular, the following "just forward back next packet" cases have been turned into proper responses:
 ResolveAddress/AddressResolution
 RegisterNode/RegisterComplete

Running without an MNC now has a very small chance of actually working.
"ROUTE OPEN" now contains data on which route is being opened (which avoids cross-talk)

Various little fixes

TYKUHN2 reviewed (dismissed):
Does not appear to interfere with GERTe transmissions. GERTi maintainers should ensure this along with other potential interference.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -26,23 +26,83 @@ if (component.isAvailable("tunnel")) then
-local iAddress = 0
+
+-- This only matters to the "GERTi" API way down below. Should be int - see startup procedure.
+-- It is updated by handler["RegisterComplete"] if the message concerns this node.
```
Please remove lines 30 and 31

20kdc replied:
These are comments explaining that iAddress is only used by the GERTi API and not by general routing.

MajorGeneralRelativity replied:
They can stay for now, but I will probably reword them

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -26,23 +26,83 @@ if (component.isAvailable("tunnel")) then
 local cachedAddress = {{}, {}, {}, {}, {}}
 local addressDex = 1
 -- Tables of neighbors and connections
+-- [x]{
```
Please change to neighbors[x]{

20kdc replied:
Will do that, along with the comment after this for connections.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -26,23 +26,83 @@ if (component.isAvailable("tunnel")) then
 local neighbors = {}
 local tier = 3
 local neighborDex = 1
+-- [x]{
```
Please change to connections[x]{

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -26,23 +26,83 @@ if (component.isAvailable("tunnel")) then
+	end)
+end
+-- Like a sleep, but it will exit early if a modem_message is received and then something happens.
+local function waitWithCancel(timeout, cancelCheck)
```
Event.pull already does this, so this entire function appears unnecessary.

20kdc replied:
It does not allow cancelling the wait *if a specific condition triggers*, which is crucial.

MajorGeneralRelativity replied:
Will review again after other requested changes are made

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -261,15 +382,54 @@ local function safedown()
+if serialTable ~= "{}" then
+	-- Even if there is no neighbor table, still register to try and form a network regardless
+	transmitInformation(neighbors[1]["address"], neighbors[1]["port"], "RegisterNode", modem.address, tier, serialTable)
+	if waitWithCancel(5, function () return iAddress end) then
```
See comment on line 90.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -278,12 +438,37 @@ function GERTi.openRoute(destination, outbound)
-		return orController(destination, modem.address, modem.address, neighbors[1]["address"], neighbors[1]["port"], neighbors[1]["port"], outbound)
+		orController(destination, modem.address, modem.address, neighbors[1]["address"], neighbors[1]["port"], neighbors[1]["port"], cb, outbound)
+	end
+	waitWithCancel(10, function () return resultGiven end)
```
See line 90.

20kdc replied:
As you can see, *cb is not nil*.

MajorGeneralRelativity replied:
Ok, my bad

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -154,45 +230,60 @@ end
-local function orController(destination, origination, beforeHop, nextHop, receivedPort, transmitPort, ...)
+-- 'cb' (callback) is given success (true) or failure (false)
+-- The ... is sent in the MNC's direction
+local function orController(destination, origination, beforeHop, nextHop, receivedPort, transmitPort, cb, ...)
```
cb is always nil, consider removing?

20kdc replied:
It, in fact, is not always nil - see GERTi.openRoute.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,87 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
 end
 
 handler["RegisterNode"] = function (eventName, receivingModem, sendingModem, port, distance, code, origination, tier, serialTable)
+	-- THIS IS COMPLETELY IGNORED. Until I have more knowledge about the organization, I won't touch this,
```
Please remove this comment

20kdc replied:
The RetrieveAddress is in fact completely ignored.

MajorGeneralRelativity replied:
Please remove this comment

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,87 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+	-- Just to make it clear, the return value system was a bad idea.
+end
+
+local function emergencyCalcIA(modemAddress)
```
Operation without MNC is not desired, and likely to cause end-user confusion. Please remove.

20kdc replied:
Understood.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,87 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+-- This is used to replace resolveAddress if the MNC is unavailable.
+-- Otherwise, this is replaced during startup procedure.
+local function addressResolver(gAddress)
+	-- A gAddress is simply a number in the MNC (unless it has a ':' in which case it goes to GERTe)
```
See line 341 comment.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,87 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
 -- transmit broadcast to check for neighboring GERTi enabled computers
 if tunnel then
 	tunnel.send("AddNeighbor")
-	receivePacket(event.pull(1, "modem_message"))
```
Unsure of why this is removed.

20kdc replied:
There is code below it that replaces it's function for both modems and tunnels.

MajorGeneralRelativity replied:
Will review after requested changes are made.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,87 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
 
 if modem then
 	modem.broadcast(4378, "AddNeighbor")
-	local continue = true
```
This does need to be reworked, but unsure of why it was removed entirely

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -261,15 +382,54 @@ local function safedown()
+		mncUnavailable = false
+	end
+end
+if mncUnavailable then
```
See comment on line 341

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -278,12 +438,37 @@ function GERTi.openRoute(destination, outbound)
+end
+
+-- Yet another low-level function
+function GERTi.resolveAddress(gAddress)
```
Please explain the value of this function.

20kdc replied:
It is impossible to effectively use openRoute externally without it. Either remove external access to openRoute or leave this be...

MajorGeneralRelativity replied:
It may be worth considering removing external access to openRoute, or perhaps folding openRoute entirely into other functionality.

20kdc commented on GERTi/GERTiClient.lua:
```diff
@@ -261,15 +382,54 @@ local function safedown()
+-- Completely new neighbors (anything that'll appear after this is over) send AddNeighbor messages,
+--  which we then forward. It's important this is high enough for all responses to be collated.
+-- The address numbers don't actually matter at all yet
+os.sleep(5)
```
*This code (line 395) here replaces 228 and 223.*

MajorGeneralRelativity replied:
Will review after changes are made.

20kdc committed:
Remove bad attempt at no-MNC support, add names to structure descriptions.

MajorGeneralRelativity commented:
Will review tonight

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -26,23 +26,83 @@ if (component.isAvailable("tunnel")) then
+		local evn, rc, sd, pt, dt, code2 = ...
+		if code ~= code2 then return end
+		if cb(...) then
+			-- Returning false from the event handler (specifically false) will get rid of the handler.
```
Please remove lines 77, 78, and 79.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -128,6 +188,22 @@ local function transmitInformation(sendTo, port, ...)
 	return false
 end
 
+-- Sends (or reuses) an existing address resolution request.
```
Unsure of what this accomplishes, as opposed to the older caching system.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
 end
 
 handler["RegisterNode"] = function (eventName, receivingModem, sendingModem, port, distance, code, origination, tier, serialTable)
+	-- THIS IS COMPLETELY IGNORED. Until I have more knowledge about the organization, I won't touch this,
```
Please remove 295 and 296.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+	storeNeighbors(eventName, receivingModem, sendingModem, port, distance, tier)
+end
+
+handler["ResolveAddress"] = function (eventName, receivingModem, sendingModem, port, distance, code, gAddress)
```
This solves the issue of this function not being included earlier. Thank you.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
-		return handler[code](eventName, receivingModem, sendingModem, port, distance, code, ...)
+		handler[code](eventName, receivingModem, sendingModem, port, distance, code, ...)
 	end
+	-- Just to make it clear, the return value system was a bad idea.
```
Please remove this comment

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
 	end
 end
 
+handler["RegisterComplete"] = function (eventName, receivingModem, sendingModem, port, distance, code, targetMA, iResponse)
```
Please remove this handler, as handler["RegisterNode"] does/will already cover this.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -26,23 +26,83 @@ if (component.isAvailable("tunnel")) then
+-- This is as opposed to addTempHandler which has no caching.
+-- Add to this with addWaitingAddress.
+-- These are cancelled after a set time.
+local waitingAddress = {}
```
Please see comments further below about difference from original caching system.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
-
---Override computer.shutdown to allow for better network leaves
+-- Override computer.shutdown to allow for better network leaves
+-- (Do this now, not later, in case we somehow are told to shutdown during sleep.
```
Please remove lines 354 and 355.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -261,15 +366,44 @@ local function safedown()
 end
 event.listen("shutdown", safedown)
+
+-- Since we're using receivePacket *anyway*, just hook ourselves in and do an os.sleep
```
Please remove 366.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -261,15 +366,44 @@ local function safedown()
+event.listen("modem_message", receivePacket)
+
+-- Wait a while to build the neighbor table.
+-- What happens here is that we get RETURNSTART messages, which we don't forward.
```
Please remove 371-374.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -261,15 +366,44 @@ local function safedown()
+local serialTable = serialize.serialize(neighbors)
+local mncUnavailable = true
+if serialTable ~= "{}" then
+	-- Even if there is no neighbor table, still register to try and form a network regardless
```
Agreed. Please remove the "if serialTable ~= "{}" then" and associated end.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -278,12 +412,37 @@ function GERTi.openRoute(destination, outbound)
+end
+
+-- Yet another low-level function
+function GERTi.resolveAddress(gAddress)
```
This function is not needed, as GERTi.openRoute will soon be removed and folded into other functionality to address complexity and space concerns.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+	-- Just to make it clear, the return value system was a bad idea.
+end
+
+local function addressResolver(gAddress)
```
Please fold in the cached check to this, as GERTi.resolveAddress no longer exists.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -330,41 +489,26 @@ end
 function GERTi.openSocket(gAddress)
-	local realAddress = nil
-	local origination = modem.address
+	local realAddress, err = GERTi.resolveAddress(gAddress)
```
Please replace with a call to local function addressResolver. See comment for line 337.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -330,41 +489,26 @@ end
+		return nil, err
 	end
-	outgoingRoute, incomingRoute, outgoingPort, incomingPort = getConnectionPair(modem.address, realAddress)
+	outgoingRoute, incomingRoute, outgoingPort, incomingPort = getConnectionPair(origination, realAddress)
```
Thank you for updating the code to be more linked card friendly.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -374,9 +518,9 @@ function GERTi.openSocket(gAddress)
 		socket.read = readData
 	else
-		print("Route cannot be opened, please confirm destination and that a valid path exists.")
+		return nil, "Route cannot be opened, please confirm destination and that a valid path exists."
```
Thank you for the better error handling.

MajorGeneralRelativity commented on GERTi/GERTiMNC.lua:
```diff
@@ -146,24 +166,35 @@ end
+                end, function () if cb then cb(false) end end)
+                return
+        end
+        sendOKResponse()
```
Why is there a sendOKResponse() at the end? Everything should have returned already, if so, this is not needed.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+	-- THIS IS COMPLETELY IGNORED. Until I have more knowledge about the organization, I won't touch this,
+	--  but FIX THIS LATER! Also see RegisterComplete and the stuff that happened to ResolveAddress - 20kdc
 	transmitInformation(neighbors[1]["address"], neighbors[1]["port"], "RetrieveAddress", origination, tier, serialTable)
 	local _, _, _, port, _, payload = event.pull(5, "modem_message")
```
I'll convert this to use the new system.

MajorGeneralRelativity commented on GERTi/GERTiMNC.lua:
```diff
@@ -230,7 +262,7 @@ handler["RegisterNode"] = function (eventName, receivingModem, sendingModem, por
 		end
 	end
-	transmitInformation(sendingModem, port, childNodes[nodeDex]["gAddress"])
+	transmitInformation(sendingModem, port, "RegisterComplete", originatorAddress, childNodes[nodeDex]["gAddress"])
```
RegisterComplete is no longer used, so these additional arguments are no longer necessary. Please remove them.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+	addWaitingAddress(gAddress, function(response)
+		-- this indentation is weird.
+		-- Anyway, this is the callback to forward the resolution
+		transmitInformation(sendingModem, port, "AddressResolution", gAddress, response)
```
Unsure of why we are switching terms here. Resolve Address should be used up the chain, and temporary handlers added to allow everything to be quickly forwarded back down.

MajorGeneralRelativity commented on GERTi/GERTiClient.lua:
```diff
@@ -201,57 +292,71 @@ handler["RemoveNeighbor"] = function (eventName, receivingModem, sendingModem, p
+	end)
+end
+
+handler["AddressResolution"] = function (eventName, receivingModem, sendingModem, port, distance, code, gAddress, mtAddress)
```
Please see line 317 as to why this should be removed.

MajorGeneralRelativity replied:
Updated comment to reference 317, not 321

MajorGeneralRelativity commented on GERTi/GERTiMNC.lua:
```diff
@@ -239,11 +271,11 @@ end
 handler["ResolveAddress"] = function (eventName, receivingModem, sendingModem, port, distance, code, gAddress)
 	if string.find(tostring(gAddress), ":") then
-		return transmitInformation(sendingModem, port, modem.address)
+		return transmitInformation(sendingModem, port, "AddressResolution", gAddress, modem.address)
```
Please see comment for line 317 of GERTiClient to explain why this needs to be changed.

MajorGeneralRelativity commented on GERTi/GERTiMNC.lua:
```diff
@@ -239,11 +271,11 @@ end
 		for key, value in pairs(childNodes) do
 			if tonumber(value["gAddress"]) == tonumber(gAddress) then
-				return value["realAddress"], transmitInformation(sendingModem, port, value["realAddress"])
+				return value["realAddress"], transmitInformation(sendingModem, port, "AddressResolution", gAddress, value["realAddress"])
```
Please see comment for line 317 of GERTiClient to explain why this needs to be changed.

MajorGeneralRelativity commented on GERTi/GERTiTestApplication.lua:
```diff
@@ -0,0 +1,29 @@
+-- This application is for use testing GERTi.
```
Thank you for adding this test application.

MajorGeneralRelativity reviewed (changes requested):
A lot of interesting changes were added. Most of them solve critical shortcomings in the GERTi protocol, and are greatly appreciated. Other changes are harder to understand, and I would appreciate a response via Discord PM, using my comments as notes. Overall, great job, but please clarify resolve some concerns.

20kdc committed:
Fix failure after success in addTempHandler

20kdc committed:
Fix some issues MGR mentioned

NOTE: DATA packets still aren't being forwarded - as far as I can tell the MNC still assumes that the node sending a registration is the node that wants to be
       registered, which is breaking things.

MajorGeneralRelativity reviewed (changes requested):
There are still changes that need to be made, or please explain why they do not need to be made.
